### PR TITLE
Improve nullable annotations for `ILogger.BeginScope`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Based on this discussion: https://github.com/dotnet/runtime/pull/66892#issuecomment-1074263709

### CODE REVIEW

**Comment by eerhardt** [comment]
cc @davidfowl @lodejard @rynowak @BrennanConroy @halter73 - FYI, in case you have any feedback on this annotation. Specifically, the `IDisposable?` return type is what I am concerned about. But we already have 2 places that can return `null` from this interface method, so it does match reality.

https://github.com/dotnet/runtime/blob/eb51b02b158c3ff71a1ec7eac8a211d1d464c1a5/src/libraries/Microsoft.Extensions.Logging.EventLog/src/EventLogLogger.cs#L51-L54

https://github.com/dotnet/aspnetcore/blob/17695facaeaf9d6d83aea9bfe87747180922117a/src/Logging.AzureAppServices/src/BatchingLogger.cs#L21-L24

**Review by eerhardt — APPROVED** [review]
LGTM - assuming no one else has any feedback.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/NullLogger.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ private NullLogger()
         }
 
         /// <inheritdoc />
-        public IDisposable BeginScope<TState>(TState state)
+        public IDisposable BeginScope<TState>(TState state) where TState : notnull
```
Comment (BrennanConroy) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/src/NullLogger.cs:23]: Just curious, how does this not complain about hiding the nullable annotation on ILogger.BeginScope?

**Review by BrennanConroy — APPROVED** [review]
Scopes aren't supported by all loggers so the `IDisposable?` makes sense to me.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/NullLogger.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/NullLogger.cs:23]: Your interface implementation can be more specific than the interface. Here the interface says "You can return an IDisposable, or null".  And this implementation says "I only ever return an IDisposable", which is allowed.

The inverse isn't allowed though.

**Review by halter73 — APPROVED** [review]

**Comment by eerhardt** [comment]
I'm going to merge this. If anyone else has feedback, we can open a new PR to address it.

**Pull request merged**